numpy/_core/src/multiarray/ctors.c

```diff
@@ -1806,26 +1806,30 @@ PyArray_CheckFromAny(PyObject *op, PyArray_Descr *descr, int min_depth,
  * Internal version of PyArray_CheckFromAny that accepts a dtypemeta. Borrows
  * references to the descriptor and dtype.
  */
-
 NPY_NO_EXPORT PyObject *
 PyArray_CheckFromAny_int(PyObject *op, PyArray_Descr *in_descr,
                          PyArray_DTypeMeta *in_DType, int min_depth,
                          int max_depth, int requires, PyObject *context)
 {
     PyObject *obj;
+    Py_XINCREF(in_descr);  /* take ownership as we may replace it */
```

Comment:
I'm a little bowled over we've gotten away with this for so long. That said, perhaps not so surprising because there aren't a lot of all sites for either `CheckFromAny` or `CheckFromAny_int`.

Can you double-check that the use of `CheckFromAny` in `iter_ass_subscript` is correct? It looks correct to me but I just want to make sure you agree.

Reply:
Just to be sure: the comment is still correct? That at the end of the day the ref-count to `in_descr` is one down?

Reply:
Sorry, I meant the comment at l.1807 for the function as a whole.

Reply:
yes, it's correct.  it borrows it, it doesn't steal it.